File: benchmarks/list_access/c/list_access.c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

typedef struct {
  int64_t px, py, pz, vx, vy, vz, mass, flags;
} Particle;

static int64_t now_ns(void) {
  struct timespec value;
  clock_gettime(CLOCK_MONOTONIC, &value);
  return (int64_t)value.tv_sec * 1000000000LL + value.tv_nsec;
}

static void report(const char* name, int64_t start, int64_t checksum) {
  printf("RESULT,c,%s,%lld,%lld\n", name,
         (long long)(now_ns() - start), (long long)checksum);
}

#define RUN_INT(NAME, BODY) do { \
  int64_t start = now_ns(), checksum = 0; \
  BODY \
  report(NAME, start, checksum); \
} while (0)

#define RUN_STRUCT(NAME, BODY) RUN_INT(NAME, BODY)

static void run_suite(const int64_t* values, const Particle* particles,
                      int64_t length, int64_t passes) {
  const int64_t operations = length * passes;
  RUN_INT("int_sequential_unchecked", {
    for (int64_t pass = 0; pass < passes; pass++)
      for (int64_t index = 0; index < length; index++) checksum += values[index];
  });
  RUN_INT("int_sequential_checked", {
    for (int64_t pass = 0; pass < passes; pass++)
      for (int64_t index = 0; index < length; index++)
        if ((uint64_t)index < (uint64_t)length) checksum += values[index];
  });
  RUN_INT("int_collection", {
    for (int64_t pass = 0; pass < passes; pass++) {
      const int64_t* end = values + length;
      for (const int64_t* value = values; value != end; value++) checksum += *value;
    }
  });
  RUN_INT("int_constant_checked", {
    volatile int64_t constant_index = 137;
    for (int64_t iteration = 0; iteration < operations; iteration++)
      if (constant_index < length) checksum += values[constant_index];
  });
  RUN_INT("int_strided_checked", {
    int64_t index = 0;
    for (int64_t iteration = 0; iteration < operations; iteration++) {
      if ((uint64_t)index < (uint64_t)length) checksum += values[index];
      index = (index + 4) % length;
    }
  });
  RUN_INT("int_random_checked", {
    for (int64_t iteration = 0; iteration < operations; iteration++) {
      int64_t index = (iteration * 48271 + 17) % length;
      if ((uint64_t)index < (uint64_t)length) checksum += values[index];
    }
  });
  RUN_INT("int_mostly_valid_checked", {
    for (int64_t iteration = 0; iteration < operations; iteration++) {
      int64_t index = iteration % 1000 == 0 ? length : (iteration * 48271 + 17) % length;
      if ((uint64_t)index < (uint64_t)length) checksum += values[index];
    }
  });
  RUN_INT("int_mixed_invalid_checked", {
    for (int64_t iteration = 0; iteration < operations; iteration++) {
      int64_t index = iteration % 4 == 0 ? -1 : (iteration * 48271 + 17) % length;
      if ((uint64_t)index < (uint64_t)length) checksum += values[index];
    }
  });
  RUN_STRUCT("struct_sequential_value", {
    for (int64_t pass = 0; pass < passes; pass++)
      for (int64_t index = 0; index < length; index++) {
        Particle particle = particles[index];
        checksum += particle.px + particle.vz + particle.mass;
      }
  });
  RUN_STRUCT("struct_sequential_pointer", {
    for (int64_t pass = 0; pass < passes; pass++)
      for (int64_t index = 0; index < length; index++) {
        const Particle* particle = &particles[index];
        checksum += particle->px + particle->vz + particle->mass;
      }
  });
  RUN_STRUCT("struct_collection_value", {
    for (int64_t pass = 0; pass < passes; pass++)
      for (int64_t index = 0; index < length; index++) {
        Particle particle = particles[index];
        checksum += particle.px + particle.vz + particle.mass;
      }
  });
  RUN_STRUCT("struct_collection_pointer", {
    for (int64_t pass = 0; pass < passes; pass++)
      for (const Particle* particle = particles; particle != particles + length; particle++)
        checksum += particle->px + particle->vz + particle->mass;
  });
  RUN_STRUCT("struct_random_pointer", {
    for (int64_t iteration = 0; iteration < operations; iteration++) {
      int64_t index = (iteration * 48271 + 17) % length;
      const Particle* particle = &particles[index];
      checksum += particle->px + particle->vz + particle->mass;
    }
  });
}

int main(void) {
  const int64_t length = 65536, passes = 128;
  int64_t* values = malloc((size_t)length * sizeof(*values));
  Particle* particles = malloc((size_t)length * sizeof(*particles));
  for (int64_t index = 0; index < length; index++) {
    values[index] = (index * 17 + 3) % 1009;
    particles[index] = (Particle){index % 97, index % 89, index % 83,
      index % 79, index % 73, index % 71, index % 67, index % 61};
  }
  for (int repetition = 0; repetition < 9; repetition++)
    run_suite(values, particles, length, passes);
  free(particles);
  free(values);
  return 0;
}

File: benchmarks/list_access/rust/list_access.rs
use std::time::Instant;

#[derive(Clone, Copy)]
struct Particle {
    px: i64, py: i64, pz: i64, vx: i64,
    vy: i64, vz: i64, mass: i64, flags: i64,
}

fn report(name: &str, start: Instant, checksum: i64) {
    println!("RESULT,rust,{},{},{}", name, start.elapsed().as_nanos(), checksum);
}

macro_rules! bench {
    ($name:expr, $checksum:ident, $body:block) => {{
        let start = Instant::now();
        let mut $checksum: i64 = 0;
        $body
        report($name, start, $checksum);
    }};
}

fn run_suite(values: &[i64], particles: &[Particle], passes: i64) {
    let length = values.len() as i64;
    let operations = length * passes;
    bench!("int_sequential_safe_index", checksum, {
        for _pass in 0..passes { for index in 0..length { checksum += values[index as usize]; } }
    });
    bench!("int_sequential_get", checksum, {
        for _pass in 0..passes { for index in 0..length { if let Some(value) = values.get(index as usize) { checksum += *value; } } }
    });
    bench!("int_iterator", checksum, {
        for _pass in 0..passes { for value in values { checksum += *value; } }
    });
    bench!("int_unchecked", checksum, {
        for _pass in 0..passes { for index in 0..length { unsafe { checksum += *values.get_unchecked(index as usize); } } }
    });
    bench!("int_constant_get", checksum, {
        let constant_index = 137_usize;
        for _iteration in 0..operations {
            let index = unsafe { std::ptr::read_volatile(&constant_index) };
            if let Some(value) = values.get(index) { checksum += *value; }
        }
    });
    bench!("int_strided_get", checksum, {
        let mut index = 0_i64;
        for _iteration in 0..operations {
            if let Some(value) = values.get(index as usize) { checksum += *value; }
            index = (index + 4) % length;
        }
    });
    bench!("int_random_get", checksum, {
        for iteration in 0..operations {
            let index = (iteration * 48271 + 17) % length;
            if let Some(value) = values.get(index as usize) { checksum += *value; }
        }
    });
    bench!("int_mostly_valid_get", checksum, {
        for iteration in 0..operations {
            let index = if iteration % 1000 == 0 { length } else { (iteration * 48271 + 17) % length };
            if let Some(value) = values.get(index as usize) { checksum += *value; }
        }
    });
    bench!("int_mixed_invalid_get", checksum, {
        for iteration in 0..operations {
            let index = if iteration % 4 == 0 { usize::MAX } else { ((iteration * 48271 + 17) % length) as usize };
            if let Some(value) = values.get(index) { checksum += *value; }
        }
    });
    bench!("struct_sequential_copy", checksum, {
        for _pass in 0..passes { for index in 0..length { let particle = particles[index as usize]; checksum += particle.px + particle.vz + particle.mass; } }
    });
    bench!("struct_sequential_get", checksum, {
        for _pass in 0..passes { for index in 0..length { if let Some(particle) = particles.get(index as usize) { checksum += particle.px + particle.vz + particle.mass; } } }
    });
    bench!("struct_iterator_copy", checksum, {
        for _pass in 0..passes { for &particle in particles { checksum += particle.px + particle.vz + particle.mass; } }
    });
    bench!("struct_iterator_ref", checksum, {
        for _pass in 0..passes { for particle in particles { checksum += particle.px + particle.vz + particle.mass; } }
    });
    bench!("struct_random_get", checksum, {
        for iteration in 0..operations {
            let index = ((iteration * 48271 + 17) % length) as usize;
            if let Some(particle) = particles.get(index) { checksum += particle.px + particle.vz + particle.mass; }
        }
    });
}

fn main() {
    let length = 65536_i64;
    let passes = 128_i64;
    let values: Vec<i64> = (0..length).map(|index| (index * 17 + 3) % 1009).collect();
    let particles: Vec<Particle> = (0..length).map(|index| Particle {
        px: index % 97, py: index % 89, pz: index % 83, vx: index % 79,
        vy: index % 73, vz: index % 71, mass: index % 67, flags: index % 61,
    }).collect();
    let field_guard = particles[0].py + particles[0].pz + particles[0].vx
        + particles[0].vy + particles[0].flags;
    std::hint::black_box(field_guard);
    for _repetition in 0..9 { run_suite(&values, &particles, passes); }
}
